src/DSN.php

```diff
@@ -0,0 +1,241 @@
+<?php
+
+/*
+ * This file is part of php-cache\adapter-bundle package.
+ *
+ * (c) 2015-2015 Aaron Scherer <aequasi@gmail.com>, Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Cache\AdapterBundle;
+
+/**
+ * @author Aaron Scherer <aequasi@gmail.com>
+ */
+class DSN
+{
+    /**
+     * @type string
+     */
+    protected $dsn;
+
+    /**
+     * @type string
+     */
+    protected $protocol;
+
+    /**
+     * @type string
+     */
+    protected $password;
+
+    /**
+     * @type string
+     */
+    protected $host;
+
+    /**
+     * @type int
+     */
+    protected $port;
+
+    /**
+     * @type string
+     */
+    protected $socket;
+
+    /**
+     * @type int
+     */
+    protected $database;
+
+    /**
+     * @type int
+     */
+    protected $weight;
+
+    /**
+     * @type string
+     */
+    protected $alias;
+
+    /**
+     * Constructor.
+     *
+     * @param string $dsn
+     */
+    public function __construct($dsn)
+    {
+        $this->dsn = $dsn;
+        $this->parseDsn($dsn);
+    }
+
+    /**
+     * @return string
+     */
+    public function getProtocol()
+    {
+        return $this->protocol;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getDatabase()
+    {
+        return $this->database;
+    }
+
+    /**
+     * @return int
+     */
+    public function getWeight()
+    {
+        return $this->weight;
+    }
+
+    /**
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getPort()
+    {
+        if (null !== $this->socket) {
+            return;
+        }
+
+        return $this->port ?: 6379;
+    }
+
+    /**
+     * @return string
+     */
+    public function getSocket()
+    {
+        return $this->socket;
+    }
+
+    /**
+     * @return string
+     */
+    public function getAlias()
+    {
+        return $this->alias;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isValid()
+    {
+        if (null === $this->getProtocol()) {
+            return false;
+        }
+
+        if (null !== $this->getHost() && null !== $this->getPort()) {
+            return true;
+        }
+
+        if (null !== $this->getSocket()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private function parseProtocol($dsn)
+    {
+        $regex = '/^(redis|mysql|mongodb|tcp):\/\//i';
```

Comment:
That would allow for any protocol to be considered valid. I only want redis, mysql, mongodb, and tcp to be considered valid for now